Add AnyObject.get_type

malept opened the pull request:
This adds a Rust port of C macros (from Ruby 2.3.1's `ruby/ruby.h`) that determines the type of a given `Value`. `AnyObject.get_type` is a helper method for this implementation.

The implementation of `rb_builtin_type` comes from https://github.com/jdm/ruby-mri-rs.

I'm not convinced that the changes are organized in the best manner, so I would be happy to move things around. Otherwise, it's working pretty well in the Rust + Ruby extension I converted from a Rust + C + Ruby hybrid.

d-unseductable commented on ruby-sys/src/util.rs:
```diff
@@ -1,7 +1,86 @@
+    SymbolFlag = 0x0c,
+}
+
+pub const SPECIAL_SHIFT: usize = 8;
```
This constant is not used outside the module, so it does not have to be public

d-unseductable commented on ruby-sys/src/util.rs:
```diff
@@ -1,7 +1,86 @@
+
+pub const SPECIAL_SHIFT: usize = 8;
+
+pub fn rb_value_is_fixnum(value: Value) -> bool {
```
Same for the this function and the following ones except `rb_type`

d-unseductable commented on src/binding/util.rs:
```diff
@@ -1,6 +1,7 @@
 use binding::global::rb_cObject;
 use types::{Argc, Id, Value};
 use ruby_sys::util::{rb_const_get, rb_funcallv, rb_intern};
+pub use ruby_sys::util::rb_type as get_type;
```
Please move `pub use` declarations below `use`s

d-unseductable commented on src/types.rs:
```diff
@@ -1,4 +1,5 @@
 use ruby_sys::types;
+pub use ruby_sys::types::ValueType;
```
Same here for `pub use`

d-unseductable commented on src/class/any_object.rs:
````diff
@@ -93,6 +94,24 @@ impl AnyObject {
+    ///
+    /// assert_eq!(any_object.get_type(), ValueType::Fixnum);
+    /// ```
+    pub fn get_type(&self) -> ValueType {
````
I would prefer to omit `get_` prefix to be consistent with current codebase

malept replied:
Right. For some reason I think I got a compile error when I named it `type`. I'll try again.

malept replied:
Yes, now I remember. `type` is a keyword. Do you have a suggestion on what else to name the method?

steveklabnik replied:
Most people use `ty` instead of `type`.

d-unseductable commented:
Hi @malept!

Thank you for this feature! There're some minor notes above, and we'll merge the PR.

Could you also give a short example of how you use `ValueTypes` in your extension? Do you convert `AnyObject`s to other types depending on the result of `get_type()`?

malept commented:
> Could you also give a short example of how you use `ValueType`s in your extension? Do you convert `AnyObject`s to other types depending on the result of `get_type()`?

Sure:

```rust
fn any_object_to_string(object: AnyObject) -> String {
  match object.get_type() {
    ValueType::Symbol => object.to::<Symbol>().to_string()
    ValueType::RString => object.to::<RString>().to_string()
    _ => object.send("to_s", vec![]).to::<RString>().to_string()
  }
}
```

malept commented:
@d-unseductable I've addressed all of the code comments in https://github.com/d-unseductable/ruru/pull/13/commits/d75d4e7290467aa375c0e8ddce0c121d27c07fa1.

d-unseductable commented:
Thank you! 👍

d-unseductable commented:
@malept, I've just bumped version to `0.7.1` and uploaded the updated crate.
